pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,147 @@
+import datetime
+import json
+import re
+
+import jsonschema
+
+from pubtools.pulplib._impl import compat_attr as attr
+from ..schema import load_schema
+from .common import InvalidDataException
+
+
+LOG = logging.getLogger("pubtools.pulplib")
+
+
+def _iso_time_now():
+    return datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
+
+
+@attr.s(kw_only=True)
+class MaintainedRepo(object):
+    """Details about the maintenance"""
+
+    message = attr.ib(default=None, type=str)
+    """Why this repository is in maintenance"""
+    owner = attr.ib(default=None, type=str)
+    """Who set this repository in maintenance mode"""
+    started = attr.ib(default=None, type=str)
+    """When did the maintenance start"""
+
+
+@attr.s(kw_only=True)
+class MaintenanceReport(object):
+    """Represent the maintenance status of pulp"""
+
+    _SCHEMA = load_schema("maintenance")
+
+    last_updated = attr.ib(default=_iso_time_now(), type=str)
+    """:class:`~datetime.datetime` in UTC when this report was last updated,
+    if it's the first time the report is created, current time is used."""
+
+    last_updated_by = attr.ib(default="Content Delivery", type=str)
+    """Person/party who updated the report last time"""
+
+    repos = attr.ib(default={}, type=dict)
+    """A dictionary contains repo_id: :class:`MaintainedRepo` pairs, indicates
+    which repositories are in maintenance mode and details.
+    If empty, then it means no repositories is in maintenance mode.
+    """
+
+    @classmethod
+    def from_data(cls, data):
+        """Create a new report with a dictionary"""
+        try:
+            jsonschema.validate(instance=data, schema=cls._SCHEMA)
+        except jsonschema.exceptions.ValidationError as error:
+            LOG.exception("%s.from_data invoked with invalid data", cls.__name__)
+            raise InvalidDataException(str(error))
+
+        maintenance = cls(
+            last_updated=data["last_updated"],
+            last_updated_by=data["last_updated_by"],
+            repos={},
+        )
+        for repo_id, details in data["repos"].items():
+            maintenance.repos[repo_id] = MaintainedRepo(**details)
+
+        return maintenance
+
+    def json(self):
+        """Output a json format report.
+
+        Returns:
+            A string contains the maintenance report in json format.
+        """
+        report = {
+            "last_updated": self.last_updated,
+            "last_updated_by": self.last_updated_by,
+            "repos": {},
+        }
+
+        for repo_id, details in self.repos.items():
+            report["repos"].update(
+                {
+                    repo_id: {
+                        "message": details.message,
+                        "owner": details.owner,
+                        "started": details.started,
+                    }
+                }
+            )
+
+        return json.dumps(report, indent=4, sort_keys=True)
+
+    def add(self, repo_ids, **kwargs):
```

Comment:
Per the suggestion to make this class immutable - I think this should rather be creating and returning a modified copy of this object instead of modifying self directly.  This helps keep code thread-safe.

Reply:
Hm, set maintenance isn't likely to be run in threads, do we still need this?

Reply:
Yeah, I think so.  Thread safety is only the most obvious reason for immutability, but there are others.

I expect in practice that it'll lead to better designed code in callers, with less side effects.  Code with side effects is generally more at risk of becoming unmaintainable, and also less likely to be idempotent (and retryable).

In any case, for this library the design choice of using immutable objects for classes representing Pulp data has already been made. We should stick with it throughout, unless there are very good reasons for an exception.

(If it's unclear whether to count this class as "Pulp data" since it's something we made up ourselves and not really a supported feature of Pulp - I don't think this should make a difference from the library user's point of view.  The API seems like it's successfully hiding implementation details here such that the user doesn't have to know or care whether maintenance mode is something built into Pulp, or emulated by our library.  That's a good thing.)

Reply:
Resolving this since the requested change was made.